Source/SpatialGDK/Private/SpatialInterop.cpp

```diff
@@ -860,16 +867,148 @@ bool USpatialInterop::IsSingletonClass(UClass* Class)
+
+	if (CanSpawnReplicatedStablyNamedActors())
+	{
+		ReserveReplicatedStablyNamedActors();
+	}
+}
+
+void USpatialInterop::UnreserveReplicatedStablyNamedActor(AActor* Actor)
+{
+	// Slightly inefficient linear search, could use a hashmap
+	USpatialActorChannel* Channel = nullptr;
+	for (auto& It : ReplicatedStablyNamedActorChannelQueue)
+	{
+		if (It->Actor == Actor)
+		{
+			Channel = It;
+			break;
+		}
+	}
+
+	if (Channel != nullptr)
+	{
+		ReplicatedStablyNamedActorChannelQueue.Remove(Channel);
+	}
+}
+
+void USpatialInterop::AddReplicatedStablyNamedActorToGSM(const FEntityId& EntityId, AActor* Actor)
+{
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPathMap = *GetEntityIdToReplicatedStablyNamedPath();
+	// If the map already has the entity id, meaning the actor was already spawned, return early
+	if (EntityIdToReplicatedStablyNamedPathMap.count(EntityId.ToSpatialEntityId()) != 0)
+	{
+		return;
+	}
+
+	std::string Path = TCHAR_TO_UTF8(*Actor->GetPathName(Actor->GetWorld()));
+	EntityIdToReplicatedStablyNamedPathMap.emplace(EntityId.ToSpatialEntityId(), Path);
+
+	improbable::unreal::GlobalStateManager::Update Update;
+	Update.set_entity_id_to_replicated_stably_named_path(EntityIdToReplicatedStablyNamedPathMap);
+
+	TSharedPtr<worker::Connection> Connection = SpatialOSInstance->GetConnection().Pin();
+	Connection->SendComponentUpdate<improbable::unreal::GlobalStateManager>(worker::EntityId((long)SpatialConstants::GLOBAL_STATE_MANAGER), Update);
+}
+
+void USpatialInterop::ReserveReplicatedStablyNamedActors()
+{
+	bCanSpawnReplicatedStablyNamedActors = true;
+	EntityIdToPathMap& EntityIdToReplicatedStablyNamedPath = *GetEntityIdToReplicatedStablyNamedPath();
+
+	for (USpatialActorChannel* Channel : ReplicatedStablyNamedActorChannelQueue)
+	{
+		std::string Path = TCHAR_TO_UTF8(*Channel->Actor->GetPathName(Channel->Actor->GetWorld()));
+
+		// Inefficient, check if we have already spawned this actor
+		bool bFound = false;
+		for (const auto& Pair : EntityIdToReplicatedStablyNamedPath)
+		{
+			if (Pair.second == Path)
+			{
+				bFound = true;
+				Channel->RegisterEntityId(FEntityId{ Pair.first });
+				break;
+			}
+		}
+		if (bFound)
+		{
+			continue;
+		}
+
+		Channel->SendReserveEntityIdRequest();
+	}
+
+	ReplicatedStablyNamedActorChannelQueue.Empty();
+}
+
+void USpatialInterop::DeleteIrrelevantReplicatedStablyNamedActors(const EntityIdToPathMap& EntityIdToReplicatedStablyNamedPath)
+{
+	for (auto& Pair : EntityIdToReplicatedStablyNamedPath)
+	{
+		FString FullPath = UTF8_TO_TCHAR(Pair.second.c_str());
+		AActor* Actor = FindObject<AActor>(GetWorld(), *FullPath);
+		worker::EntityId EntityId = Pair.first;
+
+		if (Actor != nullptr && ReplicatedStablyNamedActorTimeoutMap.Find(Actor) == nullptr)
+		{
+			FTimerHandle RetryTimer;
+			FTimerDelegate TimerCallback;
+			TimerCallback.BindLambda([Actor, EntityId, this] {
+				if (Actor != nullptr && !PackageMap->GetNetGUIDFromEntityId(EntityId).IsValid())
```

Comment:
Perhaps we should use a WeakObjPtr here. Can also validate IsPendingKill() internally.